### PR TITLE
Configurable observer client implementation 

### PR DESCRIPTION
**Pull request by eltonfss**
- Adding configurable observer clients support (RabbitMQ and REST)
- Moving auth token logic to getAuthToken method o HKDatasource class

### CODE REVIEW

**Review by marcelomachado — APPROVED** [review]

**Review by marcelomachado — APPROVED** [review]

File: datasource/observer/clients/rabbitmqobserverclient.js [hunk 1]
```diff
@@ -5,25 +5,21 @@
 
 'use strict';
 
-const ObserverClient = require ('./observerclient')
+const ObserverClient = require ('./configurableobserverclient')
```
Comment (raphael-melo) [inline datasource/observer/clients/rabbitmqobserverclient.js:8]: I'd advice against this. It makes the code less clear.

File: datasource/observer/clients/restobserverclient.js [hunk 2]
```diff
@@ -8,7 +8,7 @@
 const express        = require ('express');
 const bodyParser     = require ('body-parser');
 const request        = require ('request-promise-native');
-const ObserverClient = require ('./observerclient')
+const ObserverClient = require ('./configurableobserverclient')
```
Comment (raphael-melo) [inline datasource/observer/clients/restobserverclient.js:11]: Same as before, I'd advice against this. It makes the code less clear.

File: datasource/hkobserverfactory.js [hunk 3]
```diff
@@ -18,6 +19,26 @@ for (let key in classes)
 	clients[C.getType()] = C;
 }
 
+/**
+ * Instantiate an observer client.
+ * This method will the target HKBase which observer configuration it supports (REST or RabbitMQ)
```
Comment (raphael-melo) [inline datasource/hkobserverfactory.js:24]: The sentence seems to be wrong. Perhaps: "This method will query HKBase which observer configuration it supports (...)"?

File: datasource/observer/clients/configurableobserverclient.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+ * Licensed under The MIT License [see LICENSE for details]
+ */
+
+'use strict';
+
+const ObserverClient = require ('./observerclient');
+const request        = require("request");
+const Promisify      = require("ninja-util/promisify");
+
+class ConfigurableObserverClient extends ObserverClient
```
Comment (raphael-melo) [inline datasource/observer/clients/configurableobserverclient.js:12]: Why keep both classes?

File: datasource/hkobserverfactory.js [hunk 5]
```diff
@@ -18,6 +19,26 @@ for (let key in classes)
 	clients[C.getType()] = C;
```
Comment (raphael-melo) [inline datasource/hkobserverfactory.js:19]: ObserverClient and ConfigurableObserverClient share the same type. How to reliably know which one will be instantiate? What is the use case behind allowing "default" clients? In my mind, they seem more like abstract classes.

File: datasource/observer/clients/restobserverclient.js [hunk 6]
```diff
@@ -40,74 +39,101 @@ function setupEndpoints ()
...
 		return 'rest';
 	}
 
-	init ()
+	async init ()
 	{
+		console.info(`initializing REST observer client`);
 		return new Promise ((resolve, reject) =>
 			{
 				setupEndpoints.call (this);
```
Comment (raphael-melo) [inline datasource/observer/clients/restobserverclient.js:103]: Is there any benefit for having this method on its own and not a member of the class?

File: datasource/hkobserverfactory.js [hunk 7]
```diff
@@ -18,6 +19,26 @@ for (let key in classes)
...
 }
 
+/**
+ * Instantiate an observer client.
+ * This method will the target HKBase which observer configuration it supports (REST or RabbitMQ)
+ * and will instantiate the appropriate client with the provided configurations.
+ * @param {string} basePath HKBase basePath that is used to communicate with the server
+ * @param {Object} observerOptions Observer options that are used to initialize the observer client
+ * @param {boolean} observerOptions.isObserverService if true, the factory method will assume that it
```
[the remark this answers is not in the document]
Comment (eltonfss) [reply datasource/hkobserverfactory.js:28]: Hmm. Currently the REST service is named observer service, but it has an internal service that is named notification service (which would not show up for the clients). Therefore I believe observer service is appropriate, but I would not bother to change to Notification Service if it makes more sense (although I would also have to change the source code of the service itself to use the same name).

File: datasource/observer/clients/configurableobserverclient.js (hunk 4 above)
Comment (eltonfss) [reply datasource/observer/clients/configurableobserverclient.js:12]: I believe it makes more sense to isolate the new behavior in a specialized class.

File: datasource/hkobserverfactory.js (hunk 5 above)
Comment (eltonfss) [reply datasource/hkobserverfactory.js:19]: I've added a getType to the configurable observer client

File: datasource/observer/clients/restobserverclient.js (hunk 6 above)
Comment (eltonfss) [reply datasource/observer/clients/restobserverclient.js:103]: Not really, but I didn't see the need of changing this.

File: datasource/hkobserverfactory.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (raphael-melo) [reply datasource/hkobserverfactory.js:28]: Ok, perhaps my question is: isn't more appropriate to call the RESTFull service Notification Service rather than Observer Service? Isn't its main goal to process changes in the HKB in order to notify observers?

File: datasource/observer/clients/configurableobserverclient.js (hunk 4 above)
Comment (raphael-melo) [reply datasource/observer/clients/configurableobserverclient.js:12]: Seems fair.

File: datasource/hkobserverfactory.js (hunk 5 above)
Comment (raphael-melo) [reply datasource/hkobserverfactory.js:19]: > What is the use case behind allowing "default" clients? In my mind, they seem more like abstract classes.

Comment (eltonfss) [reply datasource/hkobserverfactory.js:19]: https://github.com/ibm-hyperknowledge/hkpy/pull/26#discussion_r762149593

File: datasource/hkobserverfactory.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (eltonfss) [reply datasource/hkobserverfactory.js:28]: I guess it would make sense, but I'm not yet convinced that observer service is a bad name for the service. 
One of the reasons this name makes sense to me because it is a service that *observes* the HKBase.

The fact that it redirects the notifications to the clients would make it reasonable to rename it to notification service, but, if I were to rename it to something that conveys the notification functionality more clearly, I would call it "NotificationFilterService" or some other long name.

Nonetheless, HKBaseObserverService feels more intuitive to me, specially because HKLib and HKPy classes are named ObserverClient. Therefore, the ObserverServer can be HKBase itself (when no ObserverConfiguration is provided or the HKBaseObserverService, otherwise).

In other words, the name HKBaseObserverService makes more sense to me because, in my understanding, what the service actually "serves" are observers, that are managed by the service. The delivery of notifications that correspond to each of these observers is only a consequence of that.
In addition, all the service endpoints are related to CRUD operations of observers, and none of them mention notifications, which could be an additional argument in favor of this name.

Does that make any sense?

File: datasource/hkobserverfactory.js (hunk 3 above)
Comment (eltonfss) [reply datasource/hkobserverfactory.js:24]: Yes, I'll correct it.

File: datasource/hkobserverfactory.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (raphael-melo) [reply datasource/hkobserverfactory.js:28]: I agree with you, perhaps the issue is with the naming of the REST service's resource. I'm not sure `Observer` is the best name; Notification Filters seems like a more apt descriptor. 

I'd have to be more familiar with the openapi in order to have a position wrt your closing arguments. 

I say we leave it at that for now.

**Review by raphael-melo — APPROVED** [review]
